src/cpp/src/visual_language/cdpruner/cdpruner.cpp

```diff
@@ -0,0 +1,377 @@
+                kernel_matrix_first = m_kernel_builder.build(visual_first_half, relevance_scores_first);
+                kernel_matrix_second = m_kernel_builder.build(visual_second_half, relevance_scores_second);
+            } else {
+                // Process entire tensor without splitting
+                ov::Tensor relevance_scores = m_relevance_calc.compute(visual_first_half, text_features);
+
+                // Build single kernel matrix using relevance scores
+                kernel_matrix_first = m_kernel_builder.build(visual_first_half, relevance_scores);
+            }
+        }
+
+        // CDPruner Step 3: Apply DPP selection
+        if (use_splitting) {
+            // Use parallel DPP selection for split processing
+            selected_tokens =
+                m_dpp_selector.select(kernel_matrix_first, kernel_matrix_second, num_tokens_to_keep, split_point);
+        } else {
+            // Direct DPP selection for single tensor processing
+            selected_tokens = m_dpp_selector.select(kernel_matrix_first, num_tokens_to_keep);
+        }
+
+        return selected_tokens;
+    } catch (const std::exception& e) {
+        OPENVINO_THROW("CDPruner::select_tokens failed: ", e.what());
+    }
+}
+
+ov::Tensor CDPruner::apply_pruning(const ov::Tensor& visual_features, const ov::Tensor& text_features, bool silent) {
+    if (m_config.pruning_ratio <= 0 || m_config.pruning_ratio >= 100)
+        return visual_features;
+    const auto& visual_shape = visual_features.get_shape();
+    const auto& text_shape = text_features.get_shape();
+    size_t batch_size = visual_shape[0];
+    size_t total_tokens = visual_shape[1];
+    size_t feature_dim = visual_shape[2];
+
+    size_t num_tokens_to_keep = static_cast<size_t>(std::round(total_tokens * (1 - m_config.pruning_ratio / 100.0)));
+
+    auto selected_tokens = select_tokens(visual_features, text_features, silent);
+    m_last_selected_tokens = selected_tokens;
+
+    // Determine actual number of selected tokens (may differ from num_tokens_to_keep due to odd->even adjustment)
+    size_t actual_selected_tokens = selected_tokens.empty() ? 0 : selected_tokens[0].size();
+
+    ov::Tensor pruned_features(visual_features.get_element_type(), {batch_size, actual_selected_tokens, feature_dim});
+
+    const float* input_data = visual_features.data<const float>();
+    float* output_data = pruned_features.data<float>();
+
+    for (size_t b = 0; b < batch_size; ++b) {
+        const auto& batch_selected = selected_tokens[b];
+
+        for (size_t t = 0; t < batch_selected.size(); ++t) {
+            size_t src_token_idx = batch_selected[t];
+
+            for (size_t f = 0; f < feature_dim; ++f) {
+                size_t src_idx = b * total_tokens * feature_dim + src_token_idx * feature_dim + f;
+                size_t dst_idx = b * actual_selected_tokens * feature_dim + t * feature_dim + f;
+                output_data[dst_idx] = input_data[src_idx];
+            }
+        }
```

Comment:
```suggestion
        for (size_t t = 0; t < batch_selected.size(); ++t) {
            size_t src_token_idx = batch_selected[t];

            size_t src_idx = b * total_tokens * feature_dim + src_token_idx * feature_dim;
            size_t dst_idx = b * actual_selected_tokens * feature_dim + t * feature_dim;
            for (size_t f = 0; f < feature_dim; ++f) {
                output_data[dst_idx+f] = input_data[src_idx+f];
            }
        }
```

Just move complex computation to outside of  {}